### PR TITLE
PCHR-1418: ToilRequestInstance

### PR DESCRIPTION
**Pull request by deb1990**
As Part of this PR `TOILRequestInstance` has been developed and also related unit tests.

**Improvements**
1. `sickness-leave-request-instance_test.js` - Test cases corrected, previously the unit tests were not failing even with false conditions.

2. `leave-request-api_test.js`- Test cases for `create`, `isValid`, `update`. Previously `jasmine.any`was used for the test cases, fixed with proper values.

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js [hunk 1]
```diff
@@ -102,6 +113,23 @@ define([
...
+        promise = LeaveRequestAPI.all(null, null, null, {}, false, 'toil');
+      });
+
+      afterEach(function () {
+        $httpBackend.flush();
+      });
+
+      it('calls the getAll() method', function () {
+        expect(LeaveRequestAPI.getAll).toHaveBeenCalled();
+        expect(LeaveRequestAPI.getAll.calls.mostRecent().args[0]).toBe('TOILRequest');
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js]: do you need `toHaveBeenCalled` if `mostRecent` is being used?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js [hunk 2]
```diff
@@ -353,10 +379,34 @@ define([
...
-          expect(LeaveRequestAPI.sendPOST).toHaveBeenCalledWith(jasmine.any(String),
-            jasmine.any(String), jasmine.any(Object));
+          expect(result.leave_request_id).toEqual(jasmine.any(String));
+        });
+      });
+    });
+
+    describe('create() for toil request', function() {
+      beforeEach(function () {
+        requestData = helper.createRandomSicknessRequest();
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js]: is the toil request same as sickness request or do some values/fields change?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js [hunk 3]
```diff
@@ -353,10 +379,34 @@ define([
...
+        spyOn(LeaveRequestAPI, 'sendPOST').and.callThrough();
+        promise = LeaveRequestAPI.create(requestData, 'toil');
+      });
+
+      afterEach(function () {
+        $httpBackend.flush();
+      });
+
+      it('call endpoint', function () {
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js]: calls endpoint

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js [hunk 4]
```diff
@@ -430,10 +478,56 @@ define([
...
+        });
+      });
+
+      describe('when called with invalid data', function () {
+        beforeEach(function () {
+          requestData = helper.createRandomSicknessRequest();
+          spyOn(LeaveRequestAPI, 'isValid').and.callFake(function (params) {
+            return $q(function (resolve, reject) {
+              reject(mockData.getNotIsValid());
+            });
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js]: `$q.reject(getNotIsValid());`

similarly at otherplaces

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js [hunk 5]
```diff
@@ -537,10 +629,58 @@ define([
...
+      it('returns updated leave request', function () {
         promise.then(function (result) {
-          expect(LeaveRequestAPI.sendPOST).toHaveBeenCalled();
-          expect(LeaveRequestAPI.sendPOST).toHaveBeenCalledWith(jasmine.any(String),
-            jasmine.any(String), jasmine.any(Object));
+          expect(result.id).toBeDefined();
+        });
+      });
+
+      describe('when does not contain id set', function () {
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js]: what is id set?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js (hunk 2 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/apis/leave-request-api_test.js]: Created `createRandomTOILRequest`

**Pull request merged**